app/src/main/java/com/kamron/pogoiv/OcrHelper.java

```diff
@@ -148,28 +148,77 @@ private double getPokemonLevelFromImg(Bitmap pokemonImage, int trainerLevel) {
      * If there was no detected upgrade cost, returns -1
      *
      * @param pokemonImage The image of the full pokemon screen
-     * @return the evolution cost, or -1 on scan failure
+     * @return the evolution cost, or -1 on no upgrade cost, or -999 on scan failure
      */
     public int getPokemonEvolutionCostFromImg(Bitmap pokemonImage) {
+        int error = -999;
         Bitmap evolutionCostImage =
                 Bitmap.createBitmap(pokemonImage, (int) (widthPixels * 0.625), (int) (heightPixels * 0.86),
                         (int) (widthPixels * 0.2), (int) (heightPixels * 0.05));
         String hash = "candyCost" + hashBitmap(evolutionCostImage);
+
+        //return cache if it exists
         String stringCacheEvoCandyCost = ocrCache.get(hash);
-        int result = -1;
-        if (stringCacheEvoCandyCost == null) { //if no cached result
-            //the dark color used for text in pogo is 76,112,114
-            evolutionCostImage = replaceColors(evolutionCostImage, 76, 112, 114, Color.WHITE, 40, false);
-            tesseract.setImage(evolutionCostImage);
-            stringCacheEvoCandyCost = tesseract.getUTF8Text().replace("S", "5").replace("s", "5");
-            try {
-                result = Integer.parseInt(stringCacheEvoCandyCost);
-            } catch (NumberFormatException e) {
-                stringCacheEvoCandyCost = "-1";
-            }
-            ocrCache.put(hash, stringCacheEvoCandyCost);
+        if (stringCacheEvoCandyCost != null) return Integer.parseInt(stringCacheEvoCandyCost);
+
+        //clean the image
+        //the dark color used for text in pogo is approximately rgb 76,112,114 if you can afford evo
+        //and the red color is rgb 255 95 100 when you cant afford the evolution
+        Bitmap evolutionCostImageCanAfford = replaceColors(evolutionCostImage, 80, 110, 110, Color.WHITE, 45, false);
+        Bitmap evolutionCostImageCannotAfford = replaceColors(evolutionCostImage, 255, 95, 100, Color.WHITE, 30,
+                false);
+
+        boolean affordIsBlank = isOnlyWhite(evolutionCostImageCanAfford);
+        boolean cannotAffordIsBlank = isOnlyWhite(evolutionCostImageCannotAfford);
+        //check if fully evolved
+        if (affordIsBlank && cannotAffordIsBlank) { //if there's no red or black text, there's no text at all.
+            ocrCache.put(hash, "-1");
+            return -1;
+        }
+
+        //use the correctly refined image (refined for red or black text)
+        if (affordIsBlank) {
+            evolutionCostImage = evolutionCostImageCannotAfford;
+        } else {
+            evolutionCostImage = evolutionCostImageCanAfford;
+        }
+
+        //If not cached or fully evolved, ocr text
+        int result;
+        tesseract.setImage(evolutionCostImage);
+        String ocrResult = fixOcrLettersToNums(tesseract.getUTF8Text());
+        try {
+            result = Integer.parseInt(ocrResult);
+            ocrCache.put(hash, ocrResult);
```

Comment:
Then we could change that line. But the caching assumes OCR is deterministic given the same input, so we should remember that we can't OCR this image, not try again to get a new failure.